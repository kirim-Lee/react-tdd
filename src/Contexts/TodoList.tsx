import { createContext, useEffect, useState } from 'react';

interface IContext {
  readonly todoList: string[];
  readonly addTodo: (toDo: string) => void;
  readonly deleteTodo: (index: number) => void;
}

const TodoListContext = createContext<IContext>({
  todoList: [],
  addTodo: () => {},
  deleteTodo: () => {},
});

interface IProps {
  children: JSX.Element | JSX.Element[];
}

const TODO_LIST = 'TodoList';

const TodoListProvider = ({ children }: IProps): JSX.Element => {
  const [todoList, setTodoList] = useState<string[]>([]);

  const addTodo = (toDo: string) => {
    if (!toDo) return;

    const newList = [...todoList, toDo];
    localStorage.setItem(TODO_LIST, JSON.stringify(newList));
    setTodoList(newList);
  };

  const deleteTodo = (index: number) => {
    let list = [...todoList];
    list.splice(index, 1);
    localStorage.setItem(TODO_LIST, JSON.stringify(list));
    setTodoList(list);
  };

  useEffect(() => {
    const list = localStorage.getItem(TODO_LIST);
    if (list) {
      setTodoList(JSON.parse(list));
    }
  }, []);

  return (
    <TodoListContext.Provider value={{ todoList, addTodo, deleteTodo }}>
      {children}
    </TodoListContext.Provider>
  );
};

export { TodoListContext, TodoListProvider };
