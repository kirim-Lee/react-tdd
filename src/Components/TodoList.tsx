import React, { useContext, useState } from 'react';
import styled from 'styled-components';
import { TodoItem } from './TodoItem';
import { TodoListContext } from '../Contexts/TodoList';

const Container = styled.div`
  min-width: 350px;
  height: 400px;
  overflow-y: scroll;
  border: 1px solid #bdbdbd;
`;

export const TodoList = () => {
  const { todoList, deleteTodo } = useContext(TodoListContext);

  return (
    <Container data-testid="toDoList">
      {todoList.map((item, index) => (
        <TodoItem
          key={item}
          id={index}
          label={item}
          onDelete={() => deleteTodo(index)}
        />
      ))}
    </Container>
  );
};
