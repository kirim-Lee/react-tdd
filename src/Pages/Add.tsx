import { useNavigate } from 'react-router-dom';
import { Container, InputContainer } from '../Components';

const Add = () => {
  const navigate = useNavigate();

  const goMain = () => {
    navigate('/', { replace: true });
  };
  return (
    <Container>
      <InputContainer onAdd={goMain} />
    </Container>
  );
};
export default Add;
