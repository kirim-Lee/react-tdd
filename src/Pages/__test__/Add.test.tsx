import {
  RouterProvider,
  createMemoryRouter,
  useLocation,
} from 'react-router-dom';
import Add from '../Add';
import { fireEvent, render, screen } from '@testing-library/react';
import { TodoListProvider } from '../../Contexts/TodoList';

const TODO_LIST = 'TodoList';

describe('Add Page', () => {
  it('렌더링 테스트', () => {
    const router = createMemoryRouter([{ path: '*', element: <Add /> }], {
      initialEntries: ['/'],
      initialIndex: 0,
    });

    const { container } = render(<RouterProvider router={router} />);

    const input = screen.getByPlaceholderText(
      '할 일을 입력해 주세요'
    ) as HTMLInputElement;
    const button = screen.getByText('추가');

    expect(input).toBeInTheDocument();
    expect(button).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('할일 추가 후 루트 url로 이동하는지 확인', () => {
    const TestComponent = () => {
      const location = useLocation();
      return (
        <TodoListProvider>
          <div>{location.pathname}</div>
          <Add />
        </TodoListProvider>
      );
    };

    const router = createMemoryRouter([
      { path: '*', element: <TestComponent /> },
    ]);

    router.navigate('/add');

    render(<RouterProvider router={router} />);

    const pathName = screen.getByText('/add');
    expect(pathName).toBeInTheDocument();

    const input = screen.getByPlaceholderText(
      '할 일을 입력해 주세요'
    ) as HTMLInputElement;
    const button = screen.getByText('추가');

    fireEvent.change(input, { target: { value: 'study react 1' } });
    fireEvent.click(button);

    expect(pathName.textContent).toBe('/');
    expect(localStorage.getItem(TODO_LIST)).toBe(
      JSON.stringify(['study react 1'])
    );
  });
});
