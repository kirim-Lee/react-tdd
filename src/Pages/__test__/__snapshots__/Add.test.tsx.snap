// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add Page 렌더링 테스트 1`] = `
<div>
  <div
    class="sc-bcXHqe jnXeEi"
  >
    <div
      class="sc-pyfCe eKYSuT"
    >
      <input
        class="sc-gKPRtg dQhuIG"
        placeholder="할 일을 입력해 주세요"
        value=""
      />
      <button
        class="sc-eDvSVe cNqITy"
      >
        <div
          class="sc-jSUZER jTTUoN"
        >
          추가
        </div>
      </button>
    </div>
  </div>
</div>
`;
